File: Lab8DemchukKI-301/Lab2DemchukKI301/Calculator.py
import math
import struct


# Function to calculate the expression 2 * x / math.sin(x) for given x
def calculate_expression(x):
    try:
        return 2 * x / math.sin(x)
    except ZeroDivisionError:
        print("Division by zero is not allowed.")
        return 0


# Function to read data from a text file and calculate y for each x
def read_text_data(filename):
    data = []
    try:
        with open(filename, 'r') as file:
            for line in file:
                x = float(line)
                y = calculate_expression(x)
                data.append((x, y))
    except FileNotFoundError:
        print(f"File {filename} not found.")
    except ValueError:
        print(f"Invalid data in {filename}.")

    return data


# Function to write data to a text file
def write_text_data(data, filename):
    try:
        with open(filename, 'w') as file:
            for x, y in data:
                file.write(f"{x}\t{y}\n")
    except:
        print(f"Error writing data to {filename}.")


# Function to read binary data from a file and calculate y for each x
def read_binary_data(filename):
    data = []
    try:
        with open(filename, 'rb') as file:
            x = file.read(8)
            while x:
                x = struct.unpack('d', x)[0]
                y = calculate_expression(x)
                data.append((x, y))
                x = file.read(8)
    except FileNotFoundError:
        print(f"File {filename} not found.")
    except struct.error:
        print(f"Invalid binary data in {filename}.")

    return data


# Function to write data to a binary file
def write_binary_data(data, filename):
    try:
        with open(filename, 'wb') as file:
            for x, y in data:
                x_bytes = struct.pack('d', x)
                file.write(x_bytes)
    except:
        print(f"Error writing data to {filename}.")

File: lab5DemchukKI-301/src/Calculator.java
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Scanner;

import static java.lang.Math.abs;
import static java.lang.Math.sin;

/**
 * The Calculator class performs calculations on a list of numbers and handles file input/output.
 */
public class Calculator {

    /**
     * Calculate the result for each element in the ArrayList and write the results to a text file.
     *
     * @param arr An ArrayList of numbers on which calculations are performed.
     */
    public void calculate(ArrayList<Double> arr) {
        for (double el : arr) {
            try {
                if (el == 0)
                    throw new ArithmeticException(String.valueOf(el));
                writeTextData(2 * el / sin(el));

            } catch (ArithmeticException e) {
                System.out.println("Division by zero is not allowed!");
                e.printStackTrace();
            }
        }
    }

    /**
     * Read data from a file and store it in an ArrayList.
     *
     * @param fileName  The name of the file to read data from.
     * @param isBinary  True if the file contains binary data, false for decimal data.
     * @return An ArrayList of read data.
     */
    public ArrayList<Double> readFile(String fileName, boolean isBinary) {
        ArrayList<Double> arr = new ArrayList<>();

        try {
            FileReader fr = new FileReader(fileName);
            Scanner sc = new Scanner(fr);

            if (isBinary) {
                while (sc.hasNext())
                    arr.add((double) Integer.parseInt(sc.nextLine(), 2));
            } else {
                while (sc.hasNext())
                    arr.add(Double.valueOf(sc.nextLine()));
            }

        } catch (IOException e) {
            e.printStackTrace();
        }

        return arr;
    }

    /**
     * Write the calculated result to a text file in both decimal and binary formats.
     *
     * @param y The calculated result to be written to the file.
     */
    public void writeTextData(double y) {
        try {
            FileWriter fw = new FileWriter("result.txt", true);
            fw.write("Decimal result: " + y + "\n");
            fw.write("Binary result: " + (y < 0 ? "-" : "+")
                    + Integer.toBinaryString(abs((int) y))
                    + "."
                    + Integer.toBinaryString(
                    Integer.parseInt(
                            String.valueOf(y % 1).substring(3, 8)
                    )
            )
                    + "\n");
            fw.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }

    }
}